M3 io fixes

eatkins opened the pull request:
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines

eatkins committed:
Update io

The new io verion removes the PathFinder <-> Glob implicit translations.
It also has a number of small bug fixes related to directory listing via
FileTreeView.

eatkins committed:
Move json formatters

Organizationally this was sloppy with the FileStamp implementation
classes split by a bunch of json formatters.

eatkins committed:
Improve cache invalidation strategy

Previously the persistent attribute map was only reset when the file
event monitor detected a change. This made it possible for the cache to
be inconsistent with the state of the file system*. To fix this, I add an
observer on the file tree repository used by the continuous build that
invalidates the cache entry for any path for which it detects a change.
Invalidating the cache does not stamp the file. That only happens either
when a task asks for the stamp for that file or when the file event
monitor reports an event and we must check if the file was updated or
not.

After this change, touching a source file will not trigger a build
unless the contents of the file actually changes.

I added a test that touches one source file in a project and updates the
content of the other. If the source file that is only ever touched ever
triggers a build, then the test fails.

* This could lead to under-compilation because ExternalHooks would not
detect that the file had been updated.

eatkins committed:
Add watchPersistFileStamps key

The persistentFileStampCache does seem to work pretty well but in case
users encounter issues, I add a boolean flag that allows the user to
turn this behavior off and always re-stamp every source file in every
task evaluation run.

eatkins committed:
Fix toString for Update

eatkins committed:
Check meta build sources before task evaluation

This commit finally fixes #241 by adding support for sbt to either
print a warning or automatically reload the project if the metabuild
sources have changed. To facilitate this, I introduce a new key,
metaBuildSourceOption which has three options:
1) IgnoreSourceChanges
2) WarnOnSourceChanges
3) ReloadOnSourceChanges

When the former is set, sbt will not check if the meta build sources
have changed. Otherwise, sbt will use the buildStructure / fileInputs to
get the ChangedFiles for the metabuild. If there are any changes, it
will either warn or reload the build depending on the value of
metaBuildSourceOption.

The mechanism for diffing the files is that I add a step to EvaluateTask
where, if the project has been loaded and
metaBuildSourceOption != IgnoreSourceChanges, we evaluate the needReload
task. If we need a reload, we return an error that indicates that a
Reload is necessary. When that error is detected, the MainLoop will
prepend "reload" to the pending commands for the state. Otherwise we
just print a warning and continue.

I benchmarked the overhead of this and it wasn't too bad. I generally
saw it taking 5-20ms to perform the check. Since this is only done once
per task evaluation run, I don't think it's a big deal. When
IgnoreSourceChanges is set, there is O(10us) overhead. If performance
does become a problem, we could add a global watch service and skip the
needReload evaluation if no files have been modified.

I removed the watchTrackMetaBuild key and made it so that the continuous
builds only track the meta build when
metaBuildSourceOption == ReloadOnSourceChanges

eatkins committed:
Fix trigger bug

I got the if condition wrong which was setting the fileInputs to have a
LastModified stamp.

eatkins committed:
Always inject input tasks

I had tried to be cute and only inject certain tasks if they're actually
used, but that made it so that dynamic tasks may not have be able to use
them.

eatkins commented:
This resolves a number of issues:
#4651 
#4650 
#4649 
#4648 
#241

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1609/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4664/head-develop-pull/238/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1609/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4664">pull/4664/head      </A> | sbt/sbt@5b49eabe4688df9c497f7121e9d6603f724e28f7
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/238">pull/238/head       </A> | sbt/io@05ff469a5db5019e6d02b5a3a8798d78e2909cc8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@faabc7712317c0693e93173edd49e5b4ccfbb671
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@6e9a8241d43fb1f2502def35ee884526b3a3b691

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4664/head&zinc=develop&io=pull/238/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1610/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4664/head-develop-pull/238/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1610/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4664">pull/4664/head      </A> | sbt/sbt@182a7b0e5f0bddb3b51409a5fc042700d768dacf
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@bbbb58e79565f8f23ab9474d061909292f1575d9
  io                | <A HREF="https://github.com/sbt/io/pull/238">pull/238/head       </A> | sbt/io@05ff469a5db5019e6d02b5a3a8798d78e2909cc8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@faabc7712317c0693e93173edd49e5b4ccfbb671
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@6e9a8241d43fb1f2502def35ee884526b3a3b691

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4664/head&zinc=develop&io=pull/238/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1611/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4664/head-develop-pull/238/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1611/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4664">pull/4664/head      </A> | sbt/sbt@182a7b0e5f0bddb3b51409a5fc042700d768dacf
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@bbbb58e79565f8f23ab9474d061909292f1575d9
  io                | <A HREF="https://github.com/sbt/io/pull/238">pull/238/head       </A> | sbt/io@05ff469a5db5019e6d02b5a3a8798d78e2909cc8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@faabc7712317c0693e93173edd49e5b4ccfbb671
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@6e9a8241d43fb1f2502def35ee884526b3a3b691

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4664/head&zinc=develop&io=pull/238/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1612/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4664/head-develop-pull/238/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1612/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4664">pull/4664/head      </A> | sbt/sbt@182a7b0e5f0bddb3b51409a5fc042700d768dacf
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@bbbb58e79565f8f23ab9474d061909292f1575d9
  io                | <A HREF="https://github.com/sbt/io/pull/238">pull/238/head       </A> | sbt/io@05ff469a5db5019e6d02b5a3a8798d78e2909cc8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@faabc7712317c0693e93173edd49e5b4ccfbb671
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@6e9a8241d43fb1f2502def35ee884526b3a3b691

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4664/head&zinc=develop&io=pull/238/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1613/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4664/head-develop-pull/238/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1613/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4664">pull/4664/head      </A> | sbt/sbt@182a7b0e5f0bddb3b51409a5fc042700d768dacf
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@bbbb58e79565f8f23ab9474d061909292f1575d9
  io                | <A HREF="https://github.com/sbt/io/pull/238">pull/238/head       </A> | sbt/io@05ff469a5db5019e6d02b5a3a8798d78e2909cc8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@faabc7712317c0693e93173edd49e5b4ccfbb671
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@6e9a8241d43fb1f2502def35ee884526b3a3b691

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4664/head&zinc=develop&io=pull/238/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eed3si9n commented on main/src/main/scala/sbt/internal/nio/CheckMetaBuild.scala:
```diff
@@ -0,0 +1,58 @@
+package sbt
```
Missing file header.

eatkins committed:
Remove unused private[sbt] key

This slipped through by mistake.

eatkins committed:
Deprecation cleanup

The main project emits a number of deprecation warnings. I've isolated
the deprecation warnings related to Watch to the DeprecatedContinuous
file. I fixed the deprecation warnings where it was straightforward to
do so. After this change, there are three non-watch related changes
emitted:

1) Defaults.scala:3760 uses the deprecated InputTask.apply. This seems
   fixable but I'm not in a hurry
2) oldLoadFailed and oldLastGrep are used by Main. I think this could
   just be fixed by removing the deprecation warnings and setting them
   private[sbt] since they will still be available in the shell.

eatkins committed:
Run more scripted tests on windows

Given that there are io differences between windows and posix systems,
we should aim to run the tests that do a lot of io on windows. There are
a few tests that don't work because of some platform specific issues so
I added a filter that excludes these tests on windows in ScriptedTests.

eatkins committed:
Support char buffered stdin on windows in continuous

I finally realized that the trick is that for non cygwin windows, the
available method on the jline wrapped input stream always returns zero.
Unlike on posix, however, the read method is interruptible which means
that we can just spin up a background thread that polls from the input
stream and writes it into a buffer.

I verified that it was no longer necessary to hit <enter> after 'r' to
rerun the continuous command on my windows vm after this change.

eed3si9n reviewed (approved):
Thanks for all the fixes!
